File: First-Edition/BookSamples/CH08/Section1p2/src/Cellular/Language/Automaton.hs
{-# LANGUAGE DataKinds      #-}
{-# LANGUAGE GADTs          #-}
{-# LANGUAGE KindSignatures #-}
{-# LANGUAGE TypeFamilies   #-}
module Cellular.Language.Automaton where

import GHC.TypeLits


-- Interfaces

-- -- Cell condition

data ICellCondition where
  CellConditionWrapper :: a -> ICellCondition

type family MkCellCondition a :: ICellCondition where
  MkCellCondition a = 'CellConditionWrapper a

-- -- State

data IState where
  StateWrapper :: a -> IState

type family MkState a :: IState where
  MkState a = 'StateWrapper a

-- -- Neighborhood

data INeighborhood where
  NeighborhoodWrapper :: a -> INeighborhood

type family MkNeighborhood a :: INeighborhood where
  MkNeighborhood a = 'NeighborhoodWrapper a

-- -- Rule

data IRule where
  RuleWrapper :: a -> IRule

type family MkRule a :: IRule where
  MkRule a = 'RuleWrapper a

-- Customizable domain model

data CustomStateTransition = StateTransition
  { cstFromState :: IState
  , cstToState   :: IState
  , cstCondition :: ICellCondition
  }

newtype DefaultState = DefState IState

data CustomStep = Step
  { defState    :: DefaultState
  , transitions :: [CustomStateTransition]
  }

File: First-Edition/BookSamples/rosetta-stone/rust/rsc2p2p6_type_level_interfaces/src/cellular/language/automaton.rs
use tl_list_lib::IInterface;
use std::marker::PhantomData;

// -- Interfaces

// -- -- Cell condition

pub struct ICellCondition;

pub struct CellConditionWrapper<T> (PhantomData::<T>);
impl<T> IInterface<ICellCondition> for CellConditionWrapper<T> {
  type Interface = ICellCondition;
}


// -- -- State

pub struct IState;

pub struct StateWrapper<T> (PhantomData::<T>);
impl<T> IInterface<IState> for StateWrapper<T> {
  type Interface = IState;
}

// -- -- Neighborhood

pub struct INeighborhood;

pub struct NeighborhoodWrapper<T> (PhantomData::<T>);
impl<T> IInterface<INeighborhood> for NeighborhoodWrapper<T> {
  type Interface = INeighborhood;
}

// -- -- Rule

pub struct IRule;

pub struct RuleWrapper<T> (PhantomData::<T>);
impl<T> IInterface<IRule> for RuleWrapper<T> {
  type Interface = IRule;
}


// -- Customizable domain model

pub struct IStateTransition;

pub struct StateTransition <
    FromState: IInterface<IState>,
    ToState: IInterface<IState>,
    Condition: IInterface<ICellCondition>>
      (PhantomData::<(FromState, ToState, Condition)>);


impl<F, T, C> IInterface<IStateTransition>
  for StateTransition<F, T, C>
  where
    F: IInterface<IState>,
    T: IInterface<IState>,
    C: IInterface<ICellCondition>
{
  type Interface = IStateTransition;
}


pub struct DefaultState <State: IInterface<IState>> (PhantomData::<State>);




// pub struct Step<DEF_STATE, TRANSITIONS: LIST>
//   (PhantomData::<DEF_STATE, TRANSITIONS>);
